server/lib/amplitude.js

```diff
@@ -0,0 +1,245 @@
+  'force-auth': GROUPS.login,
+  settings: GROUPS.settings,
+  signin: GROUPS.login,
+  signup: GROUPS.registration,
+  sms: GROUPS.sms
+};
+
+const EMAIL_TYPES = {
+  'complete-reset-password': 'reset_password',
+  'complete-signin': 'login',
+  'verify-email': 'registration'
+};
+
+const NEWSLETTER_STATES = {
+  optIn: 'subscribed',
+  optOut: 'unsubscribed',
+};
+
+const EVENTS = {
+  'flow.reset-password.submit': {
+    group: GROUPS.login,
+    event: 'forgot_submit'
+  },
+  'settings.change-password.success': {
+    group: GROUPS.settings,
+    event: 'password'
+  },
+  'settings.clients.disconnect.submit': {
+    group: GROUPS.settings,
+    event: 'disconnect_device'
+  },
+  'settings.signout.success': {
+    group: GROUPS.settings,
+    event: 'logout'
+  },
+};
+
+const FUZZY_EVENTS = new Map([
+  [ /^flow\.([\w-]+)\.engage$/, {
+    isDynamicGroup: true,
+    group: eventCategory => ENGAGE_SUBMIT_EVENT_GROUPS[eventCategory],
+    event: 'engage'
+  } ],
+  [ /^flow\.[\w-]+\.forgot-password$/, {
+    group: GROUPS.login,
+    event: 'forgot_pwd'
+  } ],
+  [ /^flow\.[\w-]+\.have-account$/, {
+    group: GROUPS.registration,
+    event: 'have_account'
+  } ],
+  [ /^flow\.([\w-]+)\.submit$/, {
+    isDynamicGroup: true,
+    group: eventCategory => ENGAGE_SUBMIT_EVENT_GROUPS[eventCategory],
+    event: 'submit'
+  } ],
+  [ /^screen\.([\w-]+)$/, {
+    isDynamicGroup: true,
+    group: eventCategory => VIEW_EVENT_GROUPS[eventCategory],
+    event: 'view'
+  } ],
+  [ /^settings\.communication-preferences\.(optIn|optOut)\.success$/, {
+    group: GROUPS.settings,
+    event: 'newsletter'
+  } ],
+  [ /^([\w-]+).verification.success$/, {
+    isDynamicGroup: true,
+    group: eventCategory => eventCategory in EMAIL_TYPES ? GROUPS.email : null,
+    event: 'click'
+  } ]
+]);
+
+const NOP = () => {};
+
+const EVENT_PROPERTIES = {
+  [GROUPS.email]: mixProperties(mapEmailType, mapService),
+  [GROUPS.login]: mixProperties(mapEntrypoint, mapService),
+  [GROUPS.registration]: mixProperties(mapEntrypoint, mapService),
+  [GROUPS.settings]: mapEntrypoint,
+  [GROUPS.sms]: NOP
+};
+
+const USER_PROPERTIES = {
+  [GROUPS.email]: NOP,
+  [GROUPS.login]: mixProperties(mapUid, mapUtmProperties),
+  [GROUPS.registration]: mixProperties(mapUid, mapUtmProperties),
+  [GROUPS.settings]: mixProperties(mapUid, mapNewsletterState),
+  [GROUPS.sms]: NOP
+};
+
+module.exports = receiveEvent;
+
+function receiveEvent (event, data) {
+  if (! event || ! data) {
+    return;
+  }
+
+  const eventType = event.type;
+  let mapping = EVENTS[eventType];
+  let eventCategory;
+
+  if (! mapping) {
+    for (const [ key, value ] of FUZZY_EVENTS.entries()) {
+      const match = key.exec(eventType);
+      if (match) {
+        mapping = value;
+        if (match.length === 2) {
+          eventCategory = match[1];
+        }
+        break;
+      }
+    }
+  }
+
+  if (mapping) {
+    let group = mapping.group;
+    if (mapping.isDynamicGroup) {
```

Comment:
Could you just check whether `mapping.group` is a function instead of using an extra property?

Reply:
Obviously I could do, yep, although I think it makes the code worse. A named boolean signals the intent clearly and is a more efficient test (this code will be executed very frequently). Why is `typeof mapping.group === 'function'` better?

Reply:
Can be refactored later if needed

Reply:
> Obviously I could do, yep, although I think it makes the code worse.

I disagree, both signal the exact same thing.

> is a more efficient test (this code will be executed very frequently)

This I'll buy.

> Why is typeof mapping.group === 'function' better?

The issue isn't with the check here, the issue is with adding extra superfluous properties to the mapping declaration.